### ISSUE TITLE
elasticsearch mapping incompatible with latest ES version

### ISSUE THREAD

**Issue by dnephin** [rtfd/readthedocs.org#980]
With elasticsearch 1.3.2

```
PUT /readthedocs/section/_mapping [status:400 request:0.005s]

[action.admin.indices.mapping.put] failed to put mappings on indices [[readthedocs-20141020010132]], type [section]
org.elasticsearch.index.mapper.MapperParsingException: 
Root type mapping not empty after parsing!
Remaining fields:   [suggest : {search_analyzer=simple, index_analyzer=simple, type=completion, payloads=true}]
```

According to http://www.elasticsearch.org/guide/en/elasticsearch/reference/current/search-suggesters-completion.html `suggest` should be under `properties`.

Other stackoverflow and github issues suggest this behaviour changed in elasticsearch 1.2.x

What version of elasticsearch is rtd using?

I suspect `readthedocs/search/indexes.py:SectionIndex.get_mapping()` will need to be updated to work with a later version.  I was able to move `suggest` into `properties` and the error went away (but I'm not sure if it's still "correct".

**Comment by carukc** [comment]
Hello,

What's strange is that the Production Server docs say to install ICU Analysis for Elasticsearch version 2.3.0:

```
# Use the correct path to the plugin executable that ships with ES.
/usr/share/elasticsearch/bin/plugin -install elasticsearch/elasticsearch-analysis-icu/2.3.0
```

According to the readme for the ICU Analysis plugin, version 2.3.0 you need to use es-1.3 .

I've got ES 1.3.0 installed and working... The ICU plugin v2.3.0 is installed and working...

but When I try to run the setup script shown here: http://docs.readthedocs.org/en/latest/rtfd.html#elastic-search-setup I am getting :

```
Traceback (most recent call last):
  File "./setup_elastic.py", line 14, in <module>
    sec.put_mapping()
  File "/home/app/readthedocs/readthedocs/search/indexes.py", line 110, in put_mapping
    self.es.indices.put_mapping(index, self._type, self.get_mapping())
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 70, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/indices.py", line 186, in put_mapping
    params=params, body=body)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 223, in perform_request
    status, raw_data = connection.perform_request(method, url, params, body, ignore=ignore)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 53, in perform_request
    self._raise_error(response.status, raw_data)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/base.py", line 82, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
elasticsearch.exceptions.TransportError: TransportError(400, u'MapperParsingException[Root type mapping not empty after parsing! Remaining fields:   [suggest : {search_analyzer=simple, index_analyzer=simple, type=completion, payloads=true}]]')
```

When I check Elastic I can see that it HAS created an index but I do not seem to be able to get celery to create entries.  It's giving me:

```
TransportError: TransportError(400, u"ElasticsearchIllegalArgumentException[Can't specify parent if no parent field has been configured]")
```

I've tried to completely remove all ES database, etc and reinstalling but no luck.  At this point I'm stuck.

What version of ES should I be using?
Is there something else that I'm doing wrong?

Thanks
Chris

**Comment by carukc** [comment]
OK... I've installed ES v 1.1 and ICU 2.1.0.... Now I'm getting some docs indexed but I'm still seeing an error durring build:

```
[22/May/2015 10:11:42] ERROR [celery.worker.job:246] Task projects.tasks.update_search[ac46fe41-386c-4275-880e-9f500cb455f1] raised exception: TransportError(400, u'{"_indices":{"readthedocs-20150522101051":{"_shards":{"total":5,"successful":0,"failed":5,"failures":[{"index":"readthedocs-20150522101051","shard":4,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":0,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":3,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":1,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":2,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"}]}}}}', {u'_indices': {u'readthedocs-20150522101051': {u'_shards': {u'successful': 0, u'failed': 5, u'total': 5, u'failures': [{u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 4}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 0}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 3}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 1}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 2}]}}}})
Traceback (most recent call last):
  File "/venv/local/lib/python2.7/site-packages/celery/task/trace.py", line 233, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/celery/task/trace.py", line 420, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/app/readthedocs/readthedocs/projects/tasks.py", line 689, in update_search
    index_search_request(version=version, page_list=page_list, commit=commit)
  File "/home/app/readthedocs/readthedocs/restapi/utils.py", line 150, in index_search_request
    page_obj.delete_document(body=delete_query)
  File "/home/app/readthedocs/readthedocs/search/indexes.py", line 165, in delete_document
    return self.es.delete_by_query(**kwargs)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 70, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/__init__.py", line 559, in delete_by_query
    params=params, body=body)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 223, in perform_request
    status, raw_data = connection.perform_request(method, url, params, body, ignore=ignore)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 53, in perform_request
    self._raise_error(response.status, raw_data)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/base.py", line 82, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
TransportError: TransportError(400, u'{"_indices":{"readthedocs-20150522101051":{"_shards":{"total":5,"successful":0,"failed":5,"failures":[{"index":"readthedocs-20150522101051","shard":4,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":0,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":3,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":1,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":2,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"}]}}}}')
[22/May/2015 10:11:42] INFO [celery.worker.job:361] Task projects.tasks.fileify[fd1c2f2f-561b-45bd-9dd6-1e2d8f735b9f] succeeded in 0.077868938446s: None
```

One step closer but still not there

Chris

**Comment by carukc** [comment]
I've also tried upgrading ES 1.1 to ES 1.3 (after running the script on v1.1)

but I'm still getting:

```
[22/May/2015 10:21:42] ERROR [celery.worker.job:246] Task projects.tasks.update_search[2ce02757-6476-4ea8-bcaa-b0f82afce0ee] raised exception: TransportError(400, u'{"_indices":{"readthedocs-20150522101051":{"_shards":{"total":5,"successful":0,"failed":5,"failures":[{"index":"readthedocs-20150522101051","shard":0,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":2,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":3,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":4,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":1,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"}]}}}}', {u'_indices': {u'readthedocs-20150522101051': {u'_shards': {u'successful': 0, u'failed': 5, u'total': 5, u'failures': [{u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 0}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 2}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 3}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 4}, {u'index': u'readthedocs-20150522101051', u'reason': u'QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]', u'shard': 1}]}}}})
Traceback (most recent call last):
  File "/venv/local/lib/python2.7/site-packages/celery/task/trace.py", line 233, in trace_task
    R = retval = fun(*args, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/celery/task/trace.py", line 420, in __protected_call__
    return self.run(*args, **kwargs)
  File "/home/app/readthedocs/readthedocs/projects/tasks.py", line 689, in update_search
    index_search_request(version=version, page_list=page_list, commit=commit)
  File "/home/app/readthedocs/readthedocs/restapi/utils.py", line 150, in index_search_request
    page_obj.delete_document(body=delete_query)
  File "/home/app/readthedocs/readthedocs/search/indexes.py", line 165, in delete_document
    return self.es.delete_by_query(**kwargs)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 70, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/__init__.py", line 559, in delete_by_query
    params=params, body=body)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 223, in perform_request
    status, raw_data = connection.perform_request(method, url, params, body, ignore=ignore)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 53, in perform_request
    self._raise_error(response.status, raw_data)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/base.py", line 82, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
TransportError: TransportError(400, u'{"_indices":{"readthedocs-20150522101051":{"_shards":{"total":5,"successful":0,"failed":5,"failures":[{"index":"readthedocs-20150522101051","shard":0,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":2,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":3,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":4,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"},{"index":"readthedocs-20150522101051","shard":1,"reason":"QueryParsingException[[readthedocs-20150522101051] request does not support [bool]]"}]}}}}')
[22/May/2015 10:22:18] INFO [core.views:469] Serving index.html for Read The Docs
```

How frustrating!

**Comment by onewheelskyward** [comment]
I'm working on updating RTD to Elasticsearch 1.5 right now, I'll post here when I get the changes merged.

**Comment by carukc** [comment]
Cool... Thanks.  Even with the 15 or 20 results that I managed to get into the index will not show.  Search just sits there attempting to load results.  Hopefully your update will fix that too.  If there is anything that I can do to help, please let me know.

**Comment by ericholscher** [comment]
Believe this should now be fixed. Just merged.

**Comment by carukc** [comment]
Cool.  Thanks!

**Comment by carukc** [comment]
Hello again... still no joy for me.  I've upgraded Elastic to 1.5.2, installed the ICU plugin 2.5.0, pulld all of your changes and rerun the init script on the rtd help pages.  Unfortunately the init script seems to produce an error:

root@e70e8212482c:/home/app/readthedocs/readthedocs# python init-elastic.py
Traceback (most recent call last):
  File "init-elastic.py", line 10, in <module>
    proj.put_mapping()
  File "/home/app/readthedocs/readthedocs/search/indexes.py", line 110, in put_mapping
    self.es.indices.put_mapping(self._type, self.get_mapping(), index)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/utils.py", line 70, in _wrapped
    return func(*args, params=params, **kwargs)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/client/indices.py", line 186, in put_mapping
    params=params, body=body)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/transport.py", line 223, in perform_request
    status, raw_data = connection.perform_request(method, url, params, body, ignore=ignore)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 53, in perform_request
    self._raise_error(response.status, raw_data)
  File "/venv/local/lib/python2.7/site-packages/elasticsearch/connection/base.py", line 82, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(status_code, error_message, additional_info)
elasticsearch.exceptions.NotFoundError: TransportError(404, u'IndexMissingException[[project] missing]')

I do not seem to be seeing the error that I had before (in the celery logs) but elastic shows that even after test building the RTD docs, only 4k was used.  Likewise when I try to search, no results are returned and the page never fully refreshes.

I'm going to try to rebuild from scratch to see where that takes me.

**Comment by carukc** [comment]
I've managed to get the init working.  I think that my virtualenv config was getting in the way but I'm not sure.  Might have been the fres clone too.  I rebuilt and the init started working.

Still no joy getting search to work though.  The builds run, I have viewable docs but no search and the elastic index size remains at 575.0B .

Is there debug somewhere that I can use to see if the build process is populating elastic?

**Comment by ericholscher** [comment]
the manage.py reindex_elasticsearch command is probably the best way to run
it directly. It uses the json files generated on build to reindex.

On Wed, May 27, 2015 at 8:47 AM, Christopher Caruk <notifications@github.com
> wrote:

> I've managed to get the init working. I think that my virtualenv config
> was getting in the way but I'm not sure. Might have been the fres clone
> too. I rebuilt and the init started working.
>
> Still no joy getting search to work though. The builds run, I have
> viewable docs but no search and the elastic index size remains at 575.0B .
>
> Is there debug somewhere that I can use to see if the build process is
> populating elastic?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rtfd/readthedocs.org/issues/980#issuecomment-105968735>
> .
>



-- 
Eric Holscher
Maker of the internet residing in Portland, Oregon
http://ericholscher.com

**Comment by carukc** [comment]
Thanks Eric,  

Elastic now shows 51 docs and an index size of 314KB.  Is there a setting that would prevent it from being created during a build?

I've also found an issue with the layout of files in the clone that I've just grabbed.

In /media/javascript there is a symbolic link that points readthedocs-doc-embed.js -> ../static/core/js/readthedocs-doc-embed.js

that file is actually in ../../readthedocs/core/static/core/js/readthedocs-doc-embed.js

This fixes the version box and gets the search box working.  I will post as an issue (if it hasn't already).

Chris
